scripts/merge-pr.sh

```diff
@@ -15,30 +20,127 @@ confirm() {
+  fi
+}
+
+GH_URL_BASE="https://api.github.com"
+
+get_pr_info() {
+  echo '[ Reading PR info ]'
+  local pr=$1
+  local pr_info_url="$GH_URL_BASE/repos/${OWNER}/${REPO}/pulls/$pr"
+  set +e
+  local pr_info
+  pr_info=$(curl -fsS "$pr_info_url")
+  if [ $? -ne 0 ]; then
+    fatal "Unable to get PR info from $pr_info_url"
+  fi
+  set -e
+  if [[ $(echo "$pr_info" | jq -r .state) == closed ]]; then
+    fatal "PR $pr is closed, will not merge"
+  fi
+  if [[ $(echo "$pr_info" | jq -r .maintainer_can_modify) == true ]]; then
+    RW_PR_REPO=1
+  else
+    warn "PR does not allow 'edits from maintainers', so it will be kept open"
+  fi
+  PR_URL=$(echo "$pr_info" | jq -r .head.repo.ssh_url)
+  PR_REMOTE_NAME=pr-$pr
+  PR_BRANCH=$(echo "$pr_info" | jq -r .head.ref)
+  PR_LOCAL_BRANCH=pr-$pr
+}
+
+fetch_pr() {
+  echo '[ Fetching PR ]'
+  git remote add $PR_REMOTE_NAME $PR_URL
+  git fetch $PR_REMOTE_NAME $PR_BRANCH
+}
+
+refresh_base_branch() {
+  git fetch $REMOTE $BRANCH
+}
+
+rebase_pr() {
+  git checkout -B $PR_LOCAL_BRANCH $PR_REMOTE_NAME/$PR_BRANCH
+  git rebase $BRANCH
+}
+
+check_is_pr_single_commit() {
+  if [[ $(git rev-list $BRANCH..$PR_LOCAL_BRANCH | wc -l) -ne 1 ]] ;then
+    confirm "PR has multiple commits, continue merging without squashing them?"
+  fi
+}
+
+confirm_pr() {
+  git log -p $BRANCH..$PR_LOCAL_BRANCH
+  confirm "Do you like this PR?"
+}
 
-echo "[Update remote and checkout branch]"
-git remote update origin
-git checkout -B $BRANCH origin/$BRANCH && git pull
+sign_pr() {
+  git commit --amend --gpg-sign --signoff
+}
 
-echo "[Rebase and squash to one commit (manual)]"
-git rebase -i origin/develop
-if [[ $(git rev-list $BRANCH..$PR_LOCAL_BRACNCH | wc -l) -ne 1 ]] ;then
-    confirm "There are multiple commits in this PR. Are you sure you wish to continue without squashing?"
-fi
+verify_pr() {
+  git show --show-signature $PR_LOCAL_BRANCH
+  confirm "Is the signature on the commit correct?"
+}
 
-echo "[Verify signature and commit (manual), update PR]"
-git show --show-signature
-git push -f
+merge_pr() {
+  # If PR is specified and can be pushed into, do it to mark PR as closed
+  if [[ -n $RW_PR_REPO ]]; then
+      git push -f $PR_REMOTE_NAME $PR_LOCAL_BRANCH:$PR_BRANCH
+  fi
+  git checkout $BRANCH
+  git merge --ff-only $PR_LOCAL_BRANCH
+  git push $REMOTE $BRANCH
+}
 
-echo "[Checkout develop and merge with same SHA]"
-git checkout develop && git pull
-git merge --ff-only $BRANCH
+cleanup() {
+  if [[ -z $DEBUG ]]; then
+      git checkout -q $BRANCH
+      git branch -q -D $PR_LOCAL_BRANCH 2>/dev/null || :
+      git remote remove $PR_REMOTE_NAME 2>/dev/null || :
+  fi
+}
 
-echo "[Push to protected develop branch]"
-git push
+run() {
+  if [[ $# -ne 1 ]] ; then
+    cat <<EOF >&2
+Requirements:
+  jq
+  curl
+Usage:
+  ./merge-pr.sh <PR-ID>
 
-echo "[Clean up remote branch]"
-git push origin --delete $BRANCH
+EOF
+    exit 2
+  fi
+  load_config
+  check_pr_prereq
+  get_pr_info "$@"
+  cleanup
```

Comment:
why is cleanup happening so early ?

Reply:
This cleanup statement is at the same place it was at originally. The intention here is to checkout develop and remove any leftover branches created in a previous failed attempt at running this script with the same PR-id.